### PR TITLE
Integration with python logging implementation

### PR DESCRIPTION
**Pull request by obrix**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [X ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

Yes this is an implementation for https://github.com/powsybl/pypowsybl/issues/142

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

This integrate java logging with logback with python logging from pypowsybl

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: java/src/main/java/com/powsybl/python/CustomAppender.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
+package com.powsybl.python;
```
Comment (geofjamg) [inline java/src/main/java/com/powsybl/python/CustomAppender.java:7]: Licence

File: java/src/main/java/com/powsybl/python/CustomAppender.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.AppenderBase;
+import ch.qos.logback.core.encoder.Encoder;
+
+import java.nio.charset.StandardCharsets;
+
+public class CustomAppender extends AppenderBase<ILoggingEvent> {
```
Comment (geofjamg) [inline java/src/main/java/com/powsybl/python/CustomAppender.java:18]: Author

File: java/src/main/java/com/powsybl/python/CustomAppender.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    private Encoder<ILoggingEvent> encoder;
+
+    @Override
+    protected void append(final ILoggingEvent e) {
+        PyPowsyblApiLib.Callback logMessage = (PyPowsyblApiLib.Callback) PyPowsyblApiLib.loggerCallback;
+        logMessage.invoke(logbackLevelToPythonLevel(e.getLevel()), CTypeUtil.toCharPtr(new String(this.getEncoder().encode(e), StandardCharsets.UTF_8)));
+    }
+
+    private int logbackLevelToPythonLevel(Level l) {
```
Comment (geofjamg) [inline java/src/main/java/com/powsybl/python/CustomAppender.java]: You should move this along with the invert method from PyPowsyblLib into a dedicated class

File: java/src/main/java/com/powsybl/python/CustomAppender.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import java.nio.charset.StandardCharsets;
+
+public class CustomAppender extends AppenderBase<ILoggingEvent> {
+
+    private Encoder<ILoggingEvent> encoder;
+
+    @Override
+    protected void append(final ILoggingEvent e) {
+        PyPowsyblApiLib.Callback logMessage = (PyPowsyblApiLib.Callback) PyPowsyblApiLib.loggerCallback;
+        logMessage.invoke(logbackLevelToPythonLevel(e.getLevel()), CTypeUtil.toCharPtr(new String(this.getEncoder().encode(e), StandardCharsets.UTF_8)));
```
Comment (geofjamg) [inline java/src/main/java/com/powsybl/python/CustomAppender.java]: This is not correct, the full log is serialized into a string, we should send to python all the log detail
 - date
 - message
 - logger

File: java/src/main/java/com/powsybl/python/PyPowsyblApiLib.java [hunk 5]
```diff
@@ -58,6 +60,10 @@
 @CContext(Directives.class)
 public final class PyPowsyblApiLib {
 
+    static boolean readConfig = true;
```
Comment (geofjamg) [inline java/src/main/java/com/powsybl/python/PyPowsyblApiLib.java]: ?

File: java/src/main/java/com/powsybl/python/PyPowsyblApiLib.java [hunk 6]
```diff
@@ -687,4 +693,43 @@ public static void addMonitoredElements(IsolateThread thread, ObjectHandle secur
     public static void freeString(IsolateThread thread, CCharPointer string, ExceptionHandlerPointer exceptionHandlerPtr) {
         doCatch(exceptionHandlerPtr, () -> UnmanagedMemory.free(string));
     }
+
+    interface Callback extends CFunctionPointer {
+        @InvokeCFunctionPointer
+        void invoke(int level, CCharPointer string);
```
Comment (geofjamg) [inline java/src/main/java/com/powsybl/python/PyPowsyblApiLib.java]: Should more detailed to really dispatch to python by providing all de individual fields

File: java/src/main/java/com/powsybl/python/PyPowsyblApiLib.java [hunk 7]
```diff
@@ -687,4 +693,43 @@ public static void addMonitoredElements(IsolateThread thread, ObjectHandle secur
...
+
+    @CEntryPoint(name = "setLogLevel")
+    public static void setLogLevel(IsolateThread thread, int logLevel, ExceptionHandlerPointer exceptionHandlerPtr) {
+        doCatch(exceptionHandlerPtr, () -> {
+            Logger rootLogger = (Logger) LoggerFactory.getLogger(org.slf4j.Logger.ROOT_LOGGER_NAME);
+            rootLogger.setLevel(pythonLevelToLogbackLevel(logLevel));
+        });
+    }
+
+    private static Level pythonLevelToLogbackLevel(int l) {
```
Comment (geofjamg) [inline java/src/main/java/com/powsybl/python/PyPowsyblApiLib.java]: See previous comment

File: pypowsybl/__init__.py [hunk 8]
```diff
@@ -31,6 +33,15 @@
     "sensitivity"
 ]
 
+#Setup a default logger that is the root logger with default log level set to ERROR
+_pypowsybl.set_logger(logging.getLogger())
+logging.getLogger().setLevel(level=logging.ERROR)
+
+def set_logger(logger: Logger) -> None:
```
Comment (geofjamg) [inline pypowsybl/__init__.py]: I don't think we should allow and bother the user to set the logger
As a first version of logging
  - log everything from Java into a "powsybl" logger (logging.getLogger('powsybl'). the problem with this approach is that we lost all Java logger infos (= java packages)
  - for a future dev, we should be able to create a python logger automatically for each Java logger

File: pypowsybl/__init__.py [hunk 9]
```diff
@@ -31,6 +33,15 @@
...
+    _pypowsybl.set_logger(logger)
+
+def get_logger() -> Logger:
+    return _pypowsybl.get_logger()
 
 def set_debug_mode(debug: bool = True) -> None:
```
Comment (geofjamg) [inline pypowsybl/__init__.py]: to remove

File: tests/test_logging.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
+#
+# Copyright (c) 2020, RTE (http://www.rte-france.com)
```
Comment (geofjamg) [inline tests/test_logging.py]: 2022

File: tests/test_logging.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+class LoggingTestCase(unittest.TestCase):
+
+    def setUp(self):
+        pp.set_config_read(False)
+
+    def test_loglevel_lf(self):
+
+        #Build a PyPow logger to handle logs and install it
+        pypowLogger = logging.getLogger("PyPow")
```
Comment (geofjamg) [inline tests/test_logging.py]: Should be done automatically

File: cpp/src/pypowsybl.h [hunk 12]
```diff
@@ -310,6 +313,42 @@ std::vector<std::vector<SeriesMetadata>> getNetworkElementCreationDataframesMeta
...
+    py::object getLogger() {
+      return this->mlogger;
+    }
+
+    bool loggerInitialized() {
+      return this->initialized;
+    }
+
+  private:
+    static CppToPythonLogger* mSingleton;
```
Comment (geofjamg) [inline cpp/src/pypowsybl.h]: we prefer _xxx convention

File: cpp/src/pypowsybl.h [hunk 13]
```diff
@@ -13,6 +13,9 @@
 #include <memory>
 #include <stdexcept>
 #include "pypowsybl-api.h"
+#include <pybind11/pybind11.h>
+
+namespace py = pybind11;
```
Comment (geofjamg) [inline cpp/src/pypowsybl.h]: Until now we tried to have a pypowsybl namespace free of pybind11 code and all python binding is in bindings.cpp

File: cpp/src/pypowsybl.cpp [hunk 14]
```diff
@@ -689,4 +697,33 @@ void createElement(pypowsybl::JavaHandle network, dataframe_array* dataframes, e
     pypowsybl::callJava<>(::createElement, network, elementType, dataframes);
 }
 
+void CppToPythonLogger::logFromJava(int level, char* pArg) {
+
+    py::gil_scoped_acquire acquire;
+    if(CppToPythonLogger::get()->loggerInitialized())
+    {
```
Comment (geofjamg) [inline cpp/src/pypowsybl.cpp]: On C++, could you try to use same style as the existing code?

File: cpp/src/pypowsybl.cpp [hunk 15]
```diff
@@ -689,4 +697,33 @@ void createElement(pypowsybl::JavaHandle network, dataframe_array* dataframes, e
     pypowsybl::callJava<>(::createElement, network, elementType, dataframes);
 }
 
+void CppToPythonLogger::logFromJava(int level, char* pArg) {
```
Comment (geofjamg) [inline cpp/src/pypowsybl.cpp]: Is it the right place for this method? A free function?

File: cpp/src/pypowsybl.cpp [hunk 16]
```diff
@@ -60,6 +60,14 @@ template<typename F, typename... ARGS>
 void callJava(F f, ARGS... args) {
     GraalVmGuard guard;
     exception_handler exc;
+
+    //Explicitly update log level on java side
+    if( CppToPythonLogger::get()->loggerInitialized() ) {
```
Comment (geofjamg) [inline cpp/src/pypowsybl.cpp]: There is no sync between "loggerInitialized" and "get". Could cause a crash if no luck?

**Review by geofjamg — CHANGES_REQUESTED** [review]

File: java/src/main/java/com/powsybl/python/CustomAppender.java (hunk 4 above)
Comment (obrix) [reply java/src/main/java/com/powsybl/python/CustomAppender.java]: You lose the formatting from logback encoder then. But can still format it from python. Is this what you have in mind ? Specifiying a custom formatter from python ?

File: java/src/main/java/com/powsybl/python/PyPowsyblApiLib.java (hunk 5 above)
Comment (obrix) [reply java/src/main/java/com/powsybl/python/PyPowsyblApiLib.java]: This was present before on the main branch but is no longer used. Probably an error from the multiple merging and rebasing...

File: cpp/src/pypowsybl.h (hunk 13 above)
Comment (obrix) [reply cpp/src/pypowsybl.h]: Let's move everything logging related in a pylogging.h / pylogging.cpp.

File: pypowsybl/__init__.py (hunk 8 above)
Comment (obrix) [reply pypowsybl/__init__.py]: Done.

**Review by geofjamg — APPROVED** [review]

**Pull request merged**